impeller/toolkit/interop/impeller.cc

```diff
@@ -0,0 +1,820 @@
+  if (!dl->IsValid()) {
+    return nullptr;
+  }
+  return dl.Leak();
+}
+
+IMPELLER_EXTERN_C
+void ImpellerDisplayListBuilderDrawDisplayList(
+    ImpellerDisplayListBuilder builder,
+    ImpellerDisplayList display_list,
+    float opacity) {
+  GetPeer(builder)->DrawDisplayList(*GetPeer(display_list), opacity);
+}
+
+IMPELLER_EXTERN_C
+ImpellerSurface ImpellerSurfaceCreateWrappedFBONew(ImpellerContext context,
+                                                   uint64_t fbo,
+                                                   ImpellerPixelFormat format,
+                                                   const ImpellerISize* size) {
+  return Surface::WrapFBO(*GetPeer(context),       //
+                          fbo,                     //
+                          ToImpellerType(format),  //
+                          ToImpellerType(*size))   //
+      .Leak();
+}
+
+IMPELLER_EXTERN_C
+void ImpellerSurfaceRetain(ImpellerSurface surface) {
+  ObjectBase::SafeRetain(surface);
+}
+
+IMPELLER_EXTERN_C
+void ImpellerSurfaceRelease(ImpellerSurface surface) {
+  ObjectBase::SafeRelease(surface);
+}
+
+IMPELLER_EXTERN_C
+bool ImpellerSurfaceDrawDisplayList(ImpellerSurface surface,
+                                    ImpellerDisplayList display_list) {
+  return GetPeer(surface)->DrawDisplayList(*GetPeer(display_list));
+}
+
+IMPELLER_EXTERN_C
+void ImpellerDisplayListBuilderDrawTexture(ImpellerDisplayListBuilder builder,
+                                           ImpellerTexture texture,
+                                           const ImpellerPoint* point,
+                                           ImpellerTextureSampling sampling,
+                                           ImpellerPaint paint) {
+  GetPeer(builder)->DrawTexture(*GetPeer(texture),            //
+                                ToImpellerType(*point),       //
+                                ToDisplayListType(sampling),  //
+                                GetPeer(paint)                //
+  );
+}
+
+IMPELLER_EXTERN_C
+void ImpellerDisplayListBuilderDrawTextureRect(
+    ImpellerDisplayListBuilder builder,
+    ImpellerTexture texture,
+    const ImpellerRect* src_rect,
+    const ImpellerRect* dst_rect,
+    ImpellerTextureSampling sampling,
+    ImpellerPaint paint) {
+  GetPeer(builder)->DrawTextureRect(*GetPeer(texture),            //
+                                    ToImpellerType(*src_rect),    //
+                                    ToImpellerType(*dst_rect),    //
+                                    ToDisplayListType(sampling),  //
+                                    GetPeer(paint)                //
+  );
+}
+
+IMPELLER_EXTERN_C
+void ImpellerColorSourceRetain(ImpellerColorSource color_source) {
+  ObjectBase::SafeRetain(color_source);
+}
+
+IMPELLER_EXTERN_C
+void ImpellerColorSourceRelease(ImpellerColorSource color_source) {
+  ObjectBase::SafeRelease(color_source);
+}
+
+static std::pair<std::vector<Color>, std::vector<Scalar>> ParseColorsAndStops(
```

Comment:
It would be preferable if these constructors didn't need to copy all of there color data. Stops at least shoudn't need it since its all floats. Colors, you could make ImpellerColor static castable to DlColor and call it good.

Reply:
I was doing that but then I ran into the issue of the colors not having a colorspace (fixed now) and the fields being swapped (RGBA vs ARGB). These subtle errors due to mismatched POD layouts freaked me out a bit and decided to copy these out. I believe these are small and callers can build and keep these around if they become expensive. So IMO its safer than reinterpret casts (for now).

Reply:
ack!